[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158062



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

IvanKobzarev committed:
DEBUG sink waits

[ghstack-poisoned]

IvanKobzarev committed:
Update on "DEBUG sink waits"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78159013).

IvanKobzarev committed:
Update on "DEBUG sink waits"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78159013).

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78159013).

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

wconstab commented on torch/_inductor/comms.py:
```diff
@@ -232,153 +231,151 @@ def accumulate_time(_snode):
     # Dicts to keep track of "next" and "previous" as double-linked structure during grouping
-    _prev: dict[BaseSchedulerNode, Optional[BaseSchedulerNode]] = {}
-    _next: dict[BaseSchedulerNode, Optional[BaseSchedulerNode]] = {}
+    _prev: dict[Optional[BaseSchedulerNode], Optional[BaseSchedulerNode]] = {}
```
what's the point of storing the 'prev' node of a None? or was this just a workaround for linter complaints?

IvanKobzarev replied:
Yes, only linter complaints.

wconstab commented on torch/_inductor/comms.py:
```diff
@@ -232,153 +231,151 @@ def accumulate_time(_snode):
+                    mem_deltas[n] = _curr_memory[n] - _curr_memory[_prev[n]]  # type: ignore[index]
+                # swap (candidate, group_head...group_tail)
+                # Before:
+                # candidate_prev -0-> candidate -1-> group_head...group_tail -2-> group_tail_next
```
not really clear to me how to read this comment and the numbering

wconstab commented on torch/_inductor/comms.py:
```diff
@@ -732,16 +826,13 @@ def is_groupable(snode):
-    log_str += grouping_log_str
+    new_snodes = _group_nodes(_head, None)
+    assert len(new_snodes) == original_snodes_num
+    new_peak_memory, curr_memory = estimate_peak_memory(
```
so, did you ever work out whether estimate_peak_memory has a bug regarding 'wait_op' or anything else?

IvanKobzarev replied:
I saw some suspicious changes in peak_memory and current_memory when I swaped wait_node with the neighbor. But did not debug it.

wconstab reviewed (approved):
lgtm. thanks!!

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78159013).

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

clee2000 commented:
@pytorchbot revert -m "sorry I want to revert something else and this is causing a merge conflict, all you should need to do is rebase and remerged" -c weird

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[simple_fsdp][inductor_collectives] rewrite reorder_collectives, sink_waits_iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78159013](https://our.internmc.facebook.com/intern/diff/D78159013)

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev commented:
@pytorchbot merge

miladm commented on torch/_inductor/comms.py:
```diff
@@ -232,153 +239,151 @@ def accumulate_time(_snode):
+                    is_grp, grp_reason = is_groupable(candidate)
+                    if is_grp:
+                        group_head = candidate
+                        group_peak_memory = max(
```
do we know the network bandwidth estimate somewhere as part of our decision criteria?
